### DISCUSSION TITLE
Good ways to use begin-data and relationships

### DISCUSSION THREAD

**Discussion by lunchboxer** [beginner-corp/begin-community#79] [Q&A]
At the end of the [README](https://github.com/smallwins/begin-data#patterns), it is said that detailed guides are coming soon. If I was a little impatient, could you give me a hint as to what a good way to navigate relationships is using @begin/data ? Since it doesn't seems like the library offers any help with this, I'd normally just use a map function on my results to retrieve the related items and add them into the original object in place of the foreign key. That seems expensive with  potentially a very large number of calls to the database. 

```javascript
// get the car companies
const companies = await data.get({ table: 'companies' })

// get the cars
let cars
const pagesOfCars = data.page({ table: 'cars' })
for await (page of pagesOfCars) {
  cars = [ ...cars, ...page ]
}

// put the two together
const companiesWithCars = companies.map(company => {
  company.cars = cars.filter(car => car.manufacturer === company.key)
  return company
})
```

I haven't tested this, but this is generally the way I was thinking of doing a join with a one-to-many relationship. Was there some particular really sensible way of doing it that whoever wrote that suggested patterns was coming soon was thinking of? many-to-many?

**Comment by brianleroux** [comment]
Hey @lunchboxer ! This library is ultimately a wrapper for doing single table queries with DynamoDB so your upper bound constraints are the same as that underlying technology. (Which is to say you have to think about how to model this data but I wouldn't worry too much about performance.) Its always good to try and keep things reasonable! But you can store fairly large objects so depending on your access patterns you can model this a number of different ways. An array of straight up objects might work. Two tables and manual joins can work too. Can you share more about the use case and how you plan to access the data?

**Reply by lunchboxer** [reply]
I am actually not looking for such specific help, just wondering if there were some secret technique that I was missing. But it could be useful to talk about a real scenario. The app I'm working on right now is a graphql api for tracking primary school student data. I have not yet decided exactly how to model the data, but I'll give you a hint as to the types and their relations. I have the most experience with RDBMS like MySQL and Postgres, so I have thought of my data as a series of highly inter-related tables. How that maps onto DynamoDB is still not something I understand well.

User: {email, name, password, role: teacher/admin}
SchoolYear: string i.e. 2020-21 as key, everything but users will belong to one either directly or through a related type
Class: { SchoolYear, teacher: User, students: [Student] }
Student: { name, birthdate }

There's probably about a dozen more, but this is one place to start. This should be pretty easy to understand. The users are basically teachers, maybe admins as well. Students all belong to a particulary class as a group each school year and through the class they are related to a teacher.

**Reply by brianleroux** [reply]
No secrets! Just a boring column store so any techniques that work there (Cassandra, Dynamo, etc) would work here too. I'd start with one table and larger objects and look to break it apart when that makes sense based on access patterns. Its cheap to iterate on this type of store so you can play around with it as your UI (and access patterns) mature.

**Reply by macdaddyaz** [reply]
> This library is ultimately a wrapper for doing single table queries with DynamoDB so your upper bound constraints are the same as that underlying technology.

Does that mean that the `table` attribute is really the Dynamo hash key, and `key` is the sort key? I'm looking to apply some [single-table design principles](https://www.sensedeep.com/blog/posts/2021/dynamodb-singletable-design.html), and trying to figure out how to do that on Begin.

**Reply by brianleroux** [reply]
sorta! the hash is actually your app id so we can secure those rows. the sort is a compound of table and key. from there you should be able to decrypt ways. generally I'd say denormalize and don't be afraid to nest.

**Reply by woutercouvaras** [reply]
Hey @brianleroux, I'm kind of tagging on to lunchboxes' question.

On [this](https://github.com/beginner-corp/begin-data#patterns) page, I see the following sentence: `Batch read by passing an Array of Objects. With these building blocks you can construct secondary indexes and joins, like one-to-many and many-to-many.`.  Is the phrase "secondary indexes" here synonymous with GSIs?  If so, how does one define them?  Which leads into my next question, which is - how does one query data based on GSIs?  If not GSIs, then whatever the mechanism for modelling many-many relationships.

I've seen in this begin data [test](https://github.com/beginner-corp/begin-data/blob/09bdde101d6c5cb5b407d0d54777725474c87d54/test/integration-test.js#L278) that a `begin` keyword is exposed (very nifty).  Are there others available too?

Thanks!
